### PR TITLE
[FIX] website: do not set a current website when fetching websites

### PR DESCRIPTION
**Pull request by younn-o**
[FIX] website: do not set a current website when fetching websites
Before this commit, the website service method fetchWebsites was not
clearly defined: it was fetching the user groups, the websites, the
current website and it was also setting the service's current website.

There is a difference between the get_current_website backend method and
the getter from the website service, introduced in [1]: the
websiteService.currentWebsite value referes to the website currently
edited in the WebsitePreview client action. From that value depend other
components visibility (like the contextual menus).

So when a fetchWebsites call was done from outside the client action,
from a PageListController with [2], it was breaking the flow by setting
a currentWebsite outside the client action and it could lead to bugged
flows:
- Go to the "Pages" list view
=> The contextual "This page" menu is not shown, it's correct
- Refresh the page
=> The "This Page" menu is shown (and broken).

To fix that, the fetchWebsites method is divided into a fetchUserGroups,
the searchRead on the websites, and another call to get_current_website,
located in the client action. It allows to load the client action iframe
with the last viewed/edited website.
Also, the menu "This Page" is shown only when there is a pageDocument at
the website service level.

[1]: https://github.com/odoo/odoo/commit/31cc10b91dc7762e23b4bde9b945be0c4ce3fe3b
[2]: https://github.com/odoo/odoo/commit/940f4ee875332dafa1f379970a7683be6b3ee606

task-2687506

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website/static/tests/website_service_mock.js
Comment (qsm-odoo) [inline addons/website/static/tests/website_service_mock.js:22]: Can you explain this change?

File: addons/website/static/src/services/website_service.js
Comment (qsm-odoo) [inline addons/website/static/src/services/website_service.js:191]: Mmh, you now do this RPC everytime even in the most common case of mono-website. But we can see what to do about that later, see other comment.

File: addons/website/static/src/client_actions/website_preview/website_preview.js
Comment (qsm-odoo) [inline addons/website/static/src/client_actions/website_preview/website_preview.js:228]: It was already like that before this PR so we can already merge this if it solves bugs and runbot issues but I wonder if calling `get_current_website` is right actually... can you remind me in which case we land on the website preview and calling `get_current_website` is necessary? Can't we always use the first website or the one received via the URM param website_id?

From my broken understanding of `get_current_website` it's supposed to be "frontend only" as there is no supposed "current website" on the backend JS side... but I may be wrong. Could you clarify everything to me? cc @rdeodoo

Comment (younn-o) [reply addons/website/static/src/client_actions/website_preview/website_preview.js:228]: Calling `get_current_website` allowed us to display the last edited/last viewed website without passing explicitly the website_id to the client action. 
For example, when switching to website 2, going to the frontend, clicking on the "edit" top left button, the client action would display website 2. But indeed maybe it would be better to be explicit for the url of this button and pass the website_id parameter.

Or, if a page is created for website 2, I think the `get_current_website` will return website 2, so clicking on the "website" module would open the client action on website 2 directly.
But again, maybe it should always be passed to the url when necessary (for example when creating a page from the list view, you would be redirected to the client action, in edit mode, with the website_id param), otherwise, always the first in the list and the user can figure it out and reorder his websites list if he prefers to land on website 2 instead of website 1.

File: addons/website/static/tests/website_service_mock.js
Comment (younn-o) [reply addons/website/static/tests/website_service_mock.js:22]: It seems that these mocks are not used anymore (after https://github.com/odoo/odoo/commit/da1a65fd2912101438655ea6b95b8eb539dbb21e). But the change is not really relevant, I will remove this diff.

Comment (qsm-odoo) [reply addons/website/static/tests/website_service_mock.js:22]: If they are not relevant anymore you can keep the change but do it in another PR with explanations so it does not confuse us during review :+1:

File: addons/website/static/src/client_actions/website_preview/website_preview.js
Comment (qsm-odoo) [reply addons/website/static/src/client_actions/website_preview/website_preview.js:228]: I guess we can keep this, I was just worried about the get_current_website actually returning something in the backend (cc @rdeodoo Is not it supposed to not return anything so the views are not cowed etc? Or do you feel it is ok relying on it?). Maybe it would be relevant for it to be in the backend session_info then? :thinking: 

But we can move forward here without this debate, @younn-o mark the PR as ready when ready and ping me to merge :+1:

Comment (rdeodoo) [reply addons/website/static/src/client_actions/website_preview/website_preview.js]: @younn-o 

> Or, if a page is created for website 2, I think the get_current_website will return website 2, so clicking on the "website" module would open the client action on website 2 directly.

Yes, but not because of the fact you created a page on website 2.
It is either due to:
- the fact you have a WIP configuration (your website don't have domains set, not a real production use case) and you used the website switcher (or any of the other flows that force a website in session)
- it simply matches always the same website, because of the URL matching

Of course, all of this can be ignored if you only have one website, whatever the flow it will always end up reading/fetching the single website.

Also, always good to remember even if not relevant here, we always develop in cases where the configuration is incorrect: we don't set any domain. Real DBs will simply jump domains and won't be able to see website 2 in website 1 URL.
That should still be true with backend > frontend merge, right?
Either you already forced a website in session somehow (eg coca.be, but shouldn't even be possible), and you type "fanta.be", the website preview will force the fanta website.
Either you switch website in the software, and it simply redirect the whole page (eg backend) to that domain URL. So if you are on coca.be and in the website preview and you click on Fanta.be, it redirects the whole page to fanta.be/@/ -> You will have to login again

@qsm-odoo 

> Is not it supposed to not return anything so the views are not cowed etc?

No, it always returns a website if it matches one (session forced, context passed, domain matched).
If nothing is found, it will fallback on the first one for the current company.

The only difference in the backend is that the website_id is not in the context, as website is added in the context only for frontend requests.

> Maybe it would be relevant for it to be in the backend session_info then?

I don't think so, it means we would need to always change that value (eg when we switch website etc, but also there is other flows forcing a website and we don't especially even know), which is not the purpose of the session_info right?

Comment (qsm-odoo) [reply addons/website/static/src/client_actions/website_preview/website_preview.js]: > >Maybe it would be relevant for it to be in the backend session_info then?

> I don't think so, it means we would need to always change that value (eg when we switch website etc, but also there is other flows forcing a website and we don't especially even know), which is not the purpose of the session_info right?

Well of course, I only suggested that in case calling `get_current_website` was right instead of making the RPC each time, do the call as a session_info value. In that case, that would be the purpose of session_info, no?

But if understand you well, calling get_current_website is not right :grimacing: ... but I am not sure to understand what the correct way is :grimacing: Right now the iframe src is always something like this: `/website/force/1?path=%2Fevent` after we potentially switch to the "right backend" (= reloading the backend with the right domain), with the website id being the one retrieved by the rpc calling `get_current_website`... it's probably ok but slowing things down for no reason (going through the /website/force controller, having the make a rpc to `get_current_website`, ...

Should we add TODO comments in this PR to investigate or would take a look at it in this PR @younn-o ?

Comment (younn-o) [reply addons/website/static/src/client_actions/website_preview/website_preview.js]: I can take a look at it, 
Besides that, I am also wondering if all the tests that are defined with `registerWebsitePreviewTour` should not have the '/web' url from python:
```py
self.start_tour(self.env['website'].get_client_action_url('/'), 'test_custom_snippet', login="admin")
```
or 
```py
self.start_tour('/@', 'test_custom_snippet', login="admin")
```
should probably be
```py
self.start_tour('/web', 'test_custom_snippet', login="admin")
```
Because it looks like the python test suite reaches the client action url twice (it starts the browser on the client action url, then launches the tour that redirects to the client action url again), therefore executing the setup/onWillStart twice (and probably starting loading the frontend twice as well).

It is visible with this PR that makes calls to two services instead of one:
```js
await this.websiteService.fetchUserGroups()
if (this.websiteService.hasMultiWebsites) {
    // At that time the first component, mounted by the first call to the client action url, 
    // from the python test suite, is destroyed, because the tour is redirecting the 
    // browser to the client action again, from the javascript.
    await this.orm.call(...)
}
```
This throws an error, as a destroyed component cannot call a service (and the test fails).

Comment (younn-o) [reply addons/website/static/src/client_actions/website_preview/website_preview.js]: I added a commit regarding https://github.com/odoo/odoo/pull/100055#discussion_r969622037
The tests are not failing with the click_all fix (which makes two async calls to service in the WebsitePreview's onWillStart instead of one before).
It seems like there is a very slight improvement performance-wise
On this branch, with /web instead of the client action url:
image
On master:
image

I don't know how legit it is to compare runbot timings like that, but it looks like the website tours are taking less time on this branch (website to website: 26 minutes=> 23 minutes, website_blog to website_twitter: 21 minutes => 20 minutes).

Comment (qsm-odoo) [reply addons/website/static/src/client_actions/website_preview/website_preview.js]: > Because it looks like the python test suite reaches the client action url twice (it starts the browser on the client action url, then launches the tour that redirects to the client action url again)

Damn, I was thinking that was indeed true with the "/@" version but should not be true with the other one because, for me, all python test should be declared with `self.env['website'].get_client_action_url` (or have some magic doing that automatically) so that it ends up on the right place directly and as the JS then ask to redirect to the /web?action=website_previewmachin, it would not do it as it is on the same URL... but that's in fact not true because the URL was automatically changed to "/@/machin" so the JS do makes a backend-leaving redirection :grimacing: 

Setting `/web` in all tests would indeed solve the problem I think as the tour on the JS side would then redirect to /web?action=machin, which does not really "reload" the page, just change the action. I would say it's not a very good practice relying on that though, the URL param on the JS side was only made to ease starting tour manually in your browser. We also have JS tours who cannot know the URL as it is generated on the python side so you would not be able to solve the problem for them.

What we should do IMO:
- All python tests should use `/@`, clearly stating where they are to be started. /@ is more fancy and easier to use and read... arguably it is a bit slower to go through it rather than calling `get_client_action_url` directly but I think it is very negligible.
- On the JS side, you normally land on `/web?action=preview_machin_etc` which is then changed quickly into `/@/machin`:
    1. Probably have to make sure that tours are not started before the URL is indeed changed into `/@/machin` (there is method in the tour system that can be patch to do that I think... even an ugly "if url is `/web?action=...`, wait and retry to check in 100ms" would be good IMO.
    2. Change the JS `getClientActionURL` function to return the `/@` version of an URL when the browser URL is not `/web=` (I almost already made that change in my previous PR but got confused): we never want to have a such a version on the client side if I am not mistaken. We either want to trigger an action change (which we can go with a owl `doAction` or something) or reach a tour URL (see point 3).
    3. Normally if 1. and 2. are done correctly the tour system won't trigger a redirect to `/@/hello` defined by a tour if you are already on `/@/hello` but if that's not working... make sure the tour system is fixed to make sure of that.

Comment (younn-o) [reply addons/website/static/src/client_actions/website_preview/website_preview.js]: I am opening another PR here to optimize the tours, optimize the queries for a user without the multi websites group, and check if we can even remove the get_current_website call for all users: https://github.com/odoo/odoo/pull/100122 
And I leave the click_all 'most primary' fix here to unblock the runbot

Comment (younn-o) [reply addons/website/static/src/client_actions/website_preview/website_preview.js]: This basic click_all fix is ready for review :)

**Review by qsm-odoo — APPROVED** [review]
As discussed, further optimization/fix will be done at https://github.com/odoo/odoo/pull/100122
This one already solves bugs and runbot issues.

@robodoo r+

**Pull request closed without merging**
